### PR TITLE
integrate go-filecoin with real PoSt generation and verification routines

### LINKED ISSUE #303: Spike: Auto update (opt in)
- Write a proposal for how to do opt-in auto update.
- See old issues from @dignifiedquire
- Pick @phritz’s brain (has experience with similar systems).
- Not much prior art here, so lean more on external research.

### LINKED ISSUE #1586: Integrate proof-of-spacetime verification
### Description

The proof-of-spacetime proof should be verified by the miner actor when someone sends it a `submitPoSt` message. This is outlined in the [spec](https://github.com/filecoin-project/specs/blob/14a7719d218ef616e40088bccd0b405b41f8b2eb/actors.md#submitpost).

In order to verify a proof-of-spacetime, the verifier needs to know how to generate the challenge seed used to generate the proof-of-spacetime, and will likely need to know the size of the sectors for which the proof-of-spacetime was generated.

### PR DESCRIPTION
**Pull request by laser**
Fixes #1586 

## What's in this PR?

1. Consume new version of rust-fil-proofs which performs real PoSts!
1. Expose bootstrap actor-boolean outside of the state machine so that bootstrap miners don't try to generate PoSts over fake commitments (created by gengen in genesis block).
1. Modify retrieval test such that it exercises seal proof verification, PoSt generation, and PoSt verification.

## Why is this PR needed?

1. We want storage miners to exercises the PoSt generation and verification code
1. We want some automated tests exercising the PoSt generation, PoSt verification, and seal proof-verification code (modifications to retrieval test)

## Changes in this rust-fil-proofs release

```

-----------------------------------------------
commit 002597027d8e720cb8010919915d704b97740e5d
Author: Erin Swenson-Healey <l@s3r.me>
Date:   Mon Mar 18 05:32:16 2019 -0700

    Feat/gf1586 integrate real post (#557)
    
    * use real PoSt
    
    * update comment
    
    * add FFI logging for seal generation and post verification/generation
    
    * error message disambiguating CommR -> FrBytes mapping issue


-----------------------------------------------
commit 7f300f01901dc97e28e54983c7e0a3d357e54376
Author: Steven Li <steven004@gmail.com>
Date:   Fri Mar 15 19:33:00 2019 +0800

    docs: fix some links
    
    - add sectorbuilder implementation and interface links
    - fix verifier links
    - remove sectorbase interfaces since they are not needed at this moment.


-----------------------------------------------
commit b28ba82f533f10e0d0fba2d2c4b187d55cc31594
Author: Friedel Ziegelmayer <dignifiedquire@users.noreply.github.com>
Date:   Thu Mar 14 20:38:22 2019 +0100

    perf(layered-drgporep): reduce memory usage in replication
    
    Reduces max resident size (maximum memory usage) during replication with zigzag (128MB  Sector)
    - from `5,925,036 KB`
    - to`3,172,848 KB`
    
    by avoiding one round of cloning graphs.


-----------------------------------------------
commit 3fad63d0ca260780e757b1a0537e49e98d9da2af
Author: Volker Mische <volker.mische@gmail.com>
Date:   Wed Mar 13 20:56:27 2019 +0100

    refactor(ffi): add a macro to free memory in the FFI automatically
    
    Dropping memory is a large part of the reposonses FFI code. With
    this new macro `#[derive(DropStructMacro)]` the `Drop` implementation
    will be auto-generated based on the struct.
    
    Example:
    
    ```rust
    pub struct FFIStagedSectorMetadata {
        pub sector_access: *const libc::c_char,
        pub sector_id: u64,
        pub pieces_len: libc::size_t,
        pub pieces_ptr: *const FFIPieceMetadata,
    
        // must be one of: Pending, Failed, Sealing
        pub seal_status_code: FFISealStatus,
    
        // if sealing failed - here's the error
        pub seal_error_msg: *const libc::c_char,
    }
    ```
    
    Will automatically create:
    
    ```rust
    impl Drop for FFIStagedSectorMetadata {
        fn drop(&mut self) {
            unsafe {
                free_c_str(self.sector_access as *mut libc::c_char);
                drop(Vec::from_raw_parts(
                    self.pieces_ptr as *mut FFIPieceMetadata,
                    self.pieces_len,
                    self.pieces_len,
                ));
                free_c_str(self.seal_error_msg as *mut libc::c_char);
            };
        }
    }
    ```
    
    To view the generated output after the macro was applied, you can use
    [cargo-expand](https://github.com/dtolnay/cargo-expand):
    
    ```console
    $ cd filecoin-proofs
    $ cargo expand --lib api::responses
    …
    ```
    
    The diff between the original code and the generated code after running
    `cargo expand --lib api::responses` and re-arranging it a bit (just
    moving the `Drop` to the same place) the diff is:
    
    ```diff
    164c164
    <                     self.faults_ptr as *mut u8,
    ---
    >                     self.faults_ptr as *mut u64,
    417c417
    <             }
    ---
    >             };
    476c476
    <             }
    ---
    >             };
    500c500
    <             }
    ---
    >             };
    530c530
    <             }
    ---
    >             };
    566c566
    <             }
    ---
    >             };
    ```
    
    Closes #303.

-----------------------------------------------
commit bb5a96e0fe198d407775cabd8b2bda35554ecfa9
Author: Lucas Molas <schomatis@gmail.com>
Date:   Fri Feb 8 00:11:55 2019 -0300

    zigzag: cache expansion parents

-----------------------------------------------
commit d3b9d27aa7fa1d00b818061b75e20351ae19e912
Author: Lucas Molas <schomatis@gmail.com>
Date:   Wed Mar 13 12:40:14 2019 -0300

    deps: lock `merkle_light` dependency to the current `master` commit
```

### CODE REVIEW

File: functional-tests/retrieval [hunk 1]
```diff
@@ -116,42 +123,68 @@ echo ""
...
-  --repodir="$CL_REPO_DIR" \
+echo "wait for commitSector sent by miner owner to be included in a block viewable by all nodes..."
+wait_for_message_in_chain_by_method_and_sender commitSector "${STORAGE_MN_MINER_OWNER_FIL_ADDR}" "${CL_REPO_DIR}"
+wait_for_message_in_chain_by_method_and_sender commitSector "${STORAGE_MN_MINER_OWNER_FIL_ADDR}" "${BOOTSTRAP_MN_REPO_DIR}"
+wait_for_message_in_chain_by_method_and_sender commitSector "${STORAGE_MN_MINER_OWNER_FIL_ADDR}" "${STORAGE_MN_REPO_DIR}"
 
 echo ""
-echo "wait for commitSector sent by miner owner to be included in a block viewable by both nodes..."
-wait_for_message_in_chain_by_method_and_sender commitSector "${MN_MINER_OWNER_FIL_ADDR}" "${CL_REPO_DIR}"
+echo "wait for for submitPoSt, too..."
```
Comment (travisperson) [inline functional-tests/retrieval]: nit: `for for`

File: protocol/storage/miner.go [hunk 2]
```diff
@@ -619,31 +621,78 @@ func (sm *Miner) currentProvingPeriodPoStChallengeSeed(ctx context.Context) (pro
 	return seed, nil
 }
 
-// OnNewHeaviestTipSet is a callback called by node, everytime the the latest head is updated.
-// It is used to check if we are in a new proving period and need to trigger PoSt submission.
-func (sm *Miner) OnNewHeaviestTipSet(ts types.TipSet) {
-	ctx := context.Background()
+// isBootstrapMinerActor is a convenience method used to determine if the miner
+// actor was created when bootstrapping the network. It it was,
```
Comment (travisperson) [inline protocol/storage/miner.go]: nit: `It it` (`If it`?), also looks like the rest of the sentence is missing.

File: protocol/storage/miner.go [hunk 3]
```diff
@@ -619,31 +621,78 @@ func (sm *Miner) currentProvingPeriodPoStChallengeSeed(ctx context.Context) (pro
...
+func (sm *Miner) isBootstrapMinerActor(ctx context.Context) (bool, error) {
+	returnValues, sig, err := sm.porcelainAPI.MessageQuery(
+		ctx,
+		address.Address{},
+		sm.minerAddr,
+		"isBootstrapMiner",
+	)
```
Comment (travisperson) [inline protocol/storage/miner.go]: Does `gengen` already add this message? I'm not too familiar with the messaging code, but I'd assume that this message needs to exist in the genesis block.

Comment (travisperson) [reply protocol/storage/miner.go]: Oh wait, I think I just understood what @phritz was saying a while back about MessageQuery and MessageSend being very similar. My guess is that this is just going to call that new method you added.

File: functional-tests/retrieval (hunk 1 above)
Comment (laser) [reply functional-tests/retrieval]: thanks you sir!

File: protocol/storage/miner.go (hunk 2 above)
Comment (laser) [reply protocol/storage/miner.go]: thanks :)

File: protocol/storage/miner.go (hunk 3 above)
Comment (laser) [reply protocol/storage/miner.go]: > My guess is that this is just going to call that new method you added.

Yes indeedly

**Pull request merged**